### PR TITLE
feat: add transaction failure reason

### PR DESCRIPTION
**Pull request by Cifko**
Description
---
Fill the transaction failure field in transaction. Add the reason to ui. Also add reason when the transaction is committed but only fee is accepted. Change transfer to still pay fees if owner tries to send too much tari.

Motivation and Context
---

How Has This Been Tested?
---
Running dan-testing. With 3 transaction (accepted, only fee is accepted, all failed). To see the ui.

What process can a PR reviewer use to test or verify this change?
---
Same as me.

Breaking Changes
---

- [x] None
- [ ] Requires data directory to be deleted
- [ ] Other - Please specify

### CODE REVIEW

File: applications/tari_dan_wallet_daemon/src/handlers/accounts.rs [hunk 1]
```diff
@@ -680,11 +681,13 @@ pub async fn handle_claim_burn(
         new_account: None,
     });
 
-    let finalized = wait_for_result(&mut events, tx_id).await?;
+    let mut finalized = wait_for_result(&mut events, tx_id).await?;
     if let Some(reject) = finalized.finalize.result.reject() {
         return Err(anyhow::anyhow!("Fee transaction rejected: {}", reject));
     }
     if let Some(reason) = finalized.transaction_failure {
+        finalized.finalize.result =
```
Comment (sdbondi) [inline applications/tari_dan_wallet_daemon/src/handlers/accounts.rs]: The transaction result types are a bit of a mess since fee instructions were added (my bad). I think that the execution engine should return this enum variant (maybe `FeeAccept` and `FullAccept` idk) rather than the wallet modifying things. Then we should be able to remove the "weird" `transaction_result` field entirely.

Comment (stringhandler) [reply applications/tari_dan_wallet_daemon/src/handlers/accounts.rs]: Yeah, I agree with that. I don't mind `FeeAccept` and `FullAccept`, but here are some other suggestions:

`FailedButCharged`
`FailedButFeesDeducted`
`OnlyFeesAccept`
`AcceptWithFailures`

File: dan_layer/engine/tests/fees.rs [hunk 2]
```diff
@@ -238,8 +238,12 @@ fn fail_partial_paid_fees() {
 
     test.disable_fees();
 
+    println!("HERE?");
```
Comment (sdbondi) [inline dan_layer/engine/tests/fees.rs]: :broom:

File: dan_layer/engine/src/transaction/processor.rs [hunk 3]
```diff
@@ -154,6 +154,14 @@ impl<TTemplateProvider: TemplateProvider<Template = LoadedTemplate> + 'static> T
                 } = runtime.interface().finalize()?;
 
                 if !fee_receipt.is_paid_in_full() {
+                    finalized.result = TransactionResult::AcceptFeeRejectRest(
+                        finalized.result.accept().unwrap().clone(),
```
Comment (sdbondi) [inline dan_layer/engine/src/transaction/processor.rs]: finalized is not guaranteed to be accept

File: integration_tests/src/wallet_daemon_cli.rs [hunk 4]
```diff
@@ -514,7 +514,7 @@ pub async fn submit_manifest_with_signing_keys(
         timeout_secs: Some(120),
     };
     let wait_resp = client.wait_transaction_result(wait_req).await.unwrap();
-    if let Some(reason) = wait_resp.transaction_failure {
+    if let Some(reason) = wait_resp.result.clone().and_then(|result| result.reject().cloned()) {
```
Comment (sdbondi) [inline integration_tests/src/wallet_daemon_cli.rs]: No need to double clone
```suggestion
    if let Some(reason) = wait_resp.result.as_ref().and_then(|result| result.reject().cloned()) {
```

File: dan_layer/wallet/sdk/src/apis/transaction.rs [hunk 5]
```diff
@@ -148,11 +146,20 @@ where
             TransactionFinalizedResult::Finalized {
                 final_decision,
                 execution_result,
-                abort_details,
+                abort_details: _,
                 json_results,
             } => {
                 let new_status = if final_decision.is_commit() {
-                    TransactionStatus::Accepted
+                    match execution_result.clone() {
```
Comment (sdbondi) [inline dan_layer/wallet/sdk/src/apis/transaction.rs]:
```suggestion
                    match execution_result.as_ref() {
```

File: applications/tari_dan_wallet_daemon/src/handlers/accounts.rs [hunk 6]
```diff
@@ -902,19 +904,21 @@ pub async fn handle_transfer(
...
             args: args![Workspace("bucket")],
         },
-        Instruction::CallMethod {
-            component_address: source_account_address,
-            method: "pay_fee".to_string(),
-            args: args![max_fee],
-        },
     ]);
 
+    fee_instructions.append(&mut vec![Instruction::CallMethod {
```
Comment (sdbondi) [inline applications/tari_dan_wallet_daemon/src/handlers/accounts.rs:909]: Why this change?

Suggest adding the instruction to the end and using `Vec::extend` rather than `Vec::append`.
```rust
// vec![....] allocates on the heap, then append takes each allocated item and the vec is dropped
    fee_instructions.append(&mut vec![Instruction::CallMethod {
```

`Vec::extend`  takes any iterator, in this case, a static array of instructions 
```suggestion
    fee_instructions.extend([Instruction::CallMethod {
```

File: integration_tests/src/wallet_daemon_cli.rs [hunk 7]
```diff
@@ -540,8 +540,8 @@ pub async fn submit_manifest(
     };
     let wait_resp = client.wait_transaction_result(wait_req).await.unwrap();
 
-    if let Some(reason) = wait_resp.transaction_failure {
-        panic!("Transaction failed: {}", reason);
+    if let Some(reason) = wait_resp.result.clone().and_then(|finalize| finalize.reject().cloned()) {
```
Comment (sdbondi) [inline integration_tests/src/wallet_daemon_cli.rs:543]:
```suggestion
    if let Some(reason) = wait_resp.result.as_ref().and_then(|finalize| finalize.reject()) {
```

File: integration_tests/src/wallet_daemon_cli.rs [hunk 8]
```diff
@@ -466,7 +466,7 @@ pub async fn submit_manifest_with_signing_keys(
         timeout_secs: Some(120),
     };
     let wait_resp = client.wait_transaction_result(wait_req).await.unwrap();
-    if let Some(reason) = wait_resp.transaction_failure {
+    if let Some(reason) = wait_resp.result.as_ref().and_then(|result| result.reject().cloned()) {
```
Comment (sdbondi) [inline integration_tests/src/wallet_daemon_cli.rs:469]:
```suggestion
    if let Some(reason) = wait_resp.result.as_ref().and_then(|result| result.reject()) {
```

File: integration_tests/src/wallet_daemon_cli.rs [hunk 9]
```diff
@@ -637,7 +637,7 @@ pub async fn create_component(
     };
     let wait_resp = client.wait_transaction_result(wait_req).await.unwrap();
 
-    if let Some(reason) = wait_resp.transaction_failure {
+    if let Some(reason) = wait_resp.result.clone().and_then(|finalize| finalize.reject().cloned()) {
```
Comment (sdbondi) [inline integration_tests/src/wallet_daemon_cli.rs:640]:
```suggestion
    if let Some(reason) = wait_resp.result.as_ref().and_then(|finalize| finalize.reject()) {
```

**Review by sdbondi — APPROVED** [review]
